proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,499 @@
+* `/path/to/dotnet-aspnet` is an alternative to `#!aspnet` that brings ASP.NET Core into scope for the file (substitute a different framework/sdk here)
+
+In effect, the `#!` directive's semantics could span a spectrum of "I need the path and command to literally execute this file" towards "I add context that affects the references available to this file".
+
+We'll need to decide if they can affect the set of references available in the file or not.
+
+### Using `#!` to point to `dotnet` or point to `env`
+
+An alternative to `#!/path/to/dotnet` is `#!/usr/bin/env dotnet`. That would mean that any UNIX program could run this file:
+
+```csharp
+#!/usr/bin/env dotnet
+WriteLine("Hello, world!");
+```
+
+However, the downside is that you can't pass any more arguments to the `dotnet` process. It's a matter of portability vs. utility.
+
+### Build output
+
+The walkthrough states that the build output is a single executable file in the current working directory. This is based on the hypothesis that this is the least surprising place to put the executable.
+
+* Is this hypothesis based on reality?
+* Is this still true if you pull in a package?
+
+### Nuget reference syntax and naming
+
+The current walkthrough proposed `"#r "nuget:..."` for packages.
+
+Syntax could change:
+
+* Use a `#!r`-style syntax, similar to the `#!` directive
+* Use a custom directive like `#nuget`
+
+Naming could change:
+
+* `package` instead of `nuget`
+
+However, that seems unlikely given that the current syntax is already shipping in .NET Interactive.
+
+### Should there be support for a user-specified TFM?
+
+So far, the assumption is that a Simple C# Program will simply execute against the TFM in the SDK it's built from. If you run from a .NET 7 SDK, it targets .NET 7. There's no way to configure that unless you use a project file.
+
+Another alternative is to use a hashbang like `dotnet-script` to support a specific TFM, like so:
+
+```csharp
+#!net5
+using System;
+
+Console.WriteLine("Hello, world!");
+```
+
+This might get tricky though, as it raises some questions:
+
+* Does the directive refer to a target runtime only?
+* Can you use features that aren't shipped with the SDK that matches a lower target runtime?
+
+This is, frankly, too complex and probably shouldn't be done. But it's worth listing.
+
+## Prior art
+
+This isn't the first proposal that deals with trying to cut down the number of concepts involved in writing a small C# program. There's some prior art to consider in this space.
+
+### A note on .NET Interactive
+
+.NET Interactive _may_ be able to service these goals, but since it is currently focused on notebooks-based programming it doesn't really address the same scenario like JavaScript, Python, and Ruby. It is highly relevant for people who are used to notebook environments, and since notebooks are breaking into educational spaces, it may also play a big role in getting people started there.
```

Comment:
.NET Interactive isn't intended just for notebooks. For example, the [`dotnet-repl`](https://github.com/jonsequitur/dotnet-repl) project allows you to run code files (including but not limited to notebook file formats) directly from the command line. I believe many of your samples will work right now using `dotnet-repl`.

One important difference currently, though, is that the semantics of C# Script differ somewhat from "classic" C#. Support for classic C# is in progress, though.